### PR TITLE
Feature: Filtering query params aka browser navigation for filtering

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

This PR introduces query parameters for document filtering. This came up a lot on -ng and as a user I often find myself wanting this i.e. when I instinctually hit the browser back button and accidentally leave the documents page. IMHO its just easier to use a keyboard shortcut to move back / forward when filtering, like undo / redo. This also has the bonus that you can directly link to a filtered view (without needing to necessarily save it).

Im pretty happy with this one! But I would def appreciate some eyes on this. The only real side-effect I see is that the last filtered view isnt saved between sessions (used to be done via `localStorage`) but that seems neglibible since you can use the browser buttons during a session and I feel that saving the last filter between sessions that are distant in time is a small use-case.

Here's some videos:

https://user-images.githubusercontent.com/4887959/160271606-0e2df6c6-eec3-492f-ad00-af7573eb6e18.mov

https://user-images.githubusercontent.com/4887959/160272000-e4c2c9b8-deab-45f0-b117-abce6633953a.mov

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain)

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://paperless-ngx.readthedocs.io/en/latest/extending.html#code-formatting-with-pre-commit-hooks).
- [ ] ~~I have made corresponding changes to the documentation as needed.~~
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

**Comment by qcasey** [comment]
This is great, bookmarking views would be very handy :)

File: src-ui/src/app/components/document-list/document-list.component.ts [hunk 1]
```diff
@@ -85,13 +91,16 @@ export class DocumentListComponent implements OnInit, OnDestroy {
...
+
     this.consumptionFinishedSubscription = this.consumerStatusService
       .onDocumentConsumptionFinished()
       .subscribe(() => {
         this.list.reload()
       })
-    this.route.paramMap.subscribe((params) => {
-      if (params.has('id')) {
+
+    this.route.paramMap
```
Comment (FrankStrieter) [inline src-ui/src/app/components/document-list/document-list.component.ts:101]: pls clear your subscription in onDestroy()

File: src-ui/src/app/components/document-list/document-list.component.ts [hunk 2]
```diff
@@ -101,11 +110,62 @@ export class DocumentListComponent implements OnInit, OnDestroy {
...
         })
-      } else {
+      })
+
+    const filterQueryVars: string[] = FILTER_RULE_TYPES.map(
+      (rt) => rt.filtervar
+    )
+
+    this.route.queryParamMap
+      .pipe(filter((qp) => !this.route.snapshot.paramMap.has('id'))) // only when not on saved view
```
Comment (FrankStrieter) [inline src-ui/src/app/components/document-list/document-list.component.ts:120]: pls clear the subscription.

File: src-ui/src/app/components/document-list/document-list.component.ts [hunk 3]
```diff
@@ -101,11 +110,62 @@ export class DocumentListComponent implements OnInit, OnDestroy {
...
+      .subscribe((queryParams) => {
+        const queryVarsFilterRules = []
+
+        filterQueryVars.forEach((filterQueryVar) => {
```
Comment (FrankStrieter) [inline src-ui/src/app/components/document-list/document-list.component.ts:124]: maybe you can extract this in a method with a clear name. It is hard to understand, what is happening here.

queryVarsFilterRules.push(`whateverIsHappeningHere(queryParams)`)

**Review by FrankStrieter — CHANGES_REQUESTED** [review]

File: src-ui/src/app/components/document-list/document-list.component.ts (hunk 3 above)
Comment (shamoon) [reply src-ui/src/app/components/document-list/document-list.component.ts]: Cheers, always appreciate your thoughts. I shrunk this logic down and hopefully made it much more readable. Its transforming query vars into "filter rules"

File: src-ui/src/app/components/document-list/document-list.component.ts [hunk 4]
```diff
@@ -59,6 +67,7 @@ export class DocumentListComponent implements OnInit, OnDestroy, AfterViewInit {
 
   unmodifiedFilterRules: FilterRule[] = []
 
+  private unsubscribeNotifier: Subject<any> = new Subject()
   private consumptionFinishedSubscription: Subscription
```
Comment (FrankStrieter) [inline src-ui/src/app/components/document-list/document-list.component.ts:71]: hey, @shamoon maybe now you can also get rid of this subscription variable.

**Review by FrankStrieter — APPROVED** [review]

File: src-ui/src/app/components/document-list/document-list.component.ts (hunk 4 above)
Comment (shamoon) [reply src-ui/src/app/components/document-list/document-list.component.ts:71]: But thats what I use to unsubscribe all? e.g. https://github.com/paperless-ngx/paperless-ngx/blob/e329d72e8b6357ab75332bb3405337664c5f79fa/src-ui/src/app/components/document-list/document-list.component.ts#L172 + https://github.com/paperless-ngx/paperless-ngx/blob/e329d72e8b6357ab75332bb3405337664c5f79fa/src-ui/src/app/components/document-list/document-list.component.ts#L192

Comment (FrankStrieter) [reply src-ui/src/app/components/document-list/document-list.component.ts:71]: I mean the `consumptionFinishedSubscription` varaiable 😄

Comment (shamoon) [reply src-ui/src/app/components/document-list/document-list.component.ts:71]: Oh haha, yes

**Pull request merged**